packages/monorepo-tools/src/where.ts

```diff
@@ -16,51 +16,58 @@
  *   npm run where "peerDependencies" -- --lerna-exec --stream --concurrency 1 -- echo "this package has peer deps"
  */
 
-const path = require('path');
-const util = require('util');
-const { runInContext, createContext } = require('vm');
-const { execFileSync } = require('child_process');
-const { forEachPackage } = require('./utils/for-each-package');
-const { MONOREPO_ROOT } = require('./utils/constants');
-
-let [expr, ...execCommandArgs] = process.argv.slice(2);
+import path from 'path';
+import util from 'util';
+import { runInContext, createContext } from 'vm';
+import { execFileSync } from 'child_process';
+import { listAllPackages } from './utils/list-all-packages';
+import { findMonorepoRoot } from './utils/find-monorepo-root';
+
+const [expr, ...execCommandArgs] = process.argv.slice(2);
 let useLernaExec = false;
 
 if (execCommandArgs.includes('--lerna-exec')) {
   execCommandArgs.splice(execCommandArgs.indexOf('--lerna-exec'), 1);
   useLernaExec = true;
 }
 
-async function filterPackagesByExpression(expression) {
-  return (
-    await forEachPackage(({ packageJson }) => {
-      try {
-        return runInContext(expression, createContext(packageJson))
-          ? packageJson.name
-          : false;
-      } catch (e) {
-        return false;
-      }
-    })
-  ).filter(Boolean);
+async function filterPackagesByExpression(
+  expression: string
+): Promise<string[]> {
+  const packages: string[] = [];
+  for await (const { packageJson } of listAllPackages()) {
+    try {
+      if (runInContext(expression, createContext(packageJson)))
+        packages.push(packageJson.name);
+    } catch {
+      /* skip */
+    }
+  }
+  return packages;
 }
 
-async function lernaExec(packages) {
+async function lernaExec(packages: string[]) {
   const scope = packages.map((name) => `--scope=${name}`);
 
   if (scope.length === 0) {
     console.info(`No packages matched filter "${expr}"`);
     return;
   }
 
-  const lernaBin = path.resolve(MONOREPO_ROOT, 'node_modules', '.bin', 'lerna');
+  const lernaBin = path.resolve(
+    await findMonorepoRoot(),
+    'node_modules',
+    '.bin',
+    'lerna'
+  );
 
   execFileSync(lernaBin, ['exec', ...scope, ...execCommandArgs], {
     stdio: 'inherit',
   });
 }
 
-async function npmWorkspaces(packages) {
+// eslint-disable-next-line @typescript-eslint/require-await
+async function npmWorkspaces(packages: string[]) {
```

Comment:
We could remove it, yes. The function interface is supposed to match the one of `lernaExec`, though, and I’d like to keep the `async` so that we could get rid of the `execFileSync` calls more easily in the future (which I think we should, I just didn’t want to overload this PR with even more changes).

Reply:
Sounds good, thanks for the explanation.